dbmongo/lib/engine/public.go

```diff
@@ -28,18 +28,20 @@ func PublicOne(batch AdminBatch, key string) error {
 		return err
 	}
 
+	jsParams := bson.M{
+		"date_debut":      batch.Params.DateDebut,
+		"date_fin":        batch.Params.DateFin,
+		"serie_periode":   misc.GenereSeriePeriode(batch.Params.DateDebut, batch.Params.DateFin),
+		"offset_effectif": (batch.Params.DateFinEffectif.Year()-batch.Params.DateFin.Year())*12 + int(batch.Params.DateFinEffectif.Month()-batch.Params.DateFin.Month()),
+		"actual_batch":    batch.ID.Key,
+		"naf":             naf,
+		"batches":         GetBatchesID(),
+		"types":           GetTypes(),
+	}
+
 	scope := bson.M{
-		"date_debut":             batch.Params.DateDebut,
-		"date_fin":               batch.Params.DateFin,
-		"date_fin_effectif":      batch.Params.DateFinEffectif,
-		"serie_periode":          misc.GenereSeriePeriode(batch.Params.DateDebut, batch.Params.DateFin),
-		"serie_periode_annuelle": misc.GenereSeriePeriodeAnnuelle(batch.Params.DateDebut, batch.Params.DateFin),
```

Comment:
je vois que cette variable n'a pas été reprise, elle n'était jamais utilisée ? (c'est possible…)
Cela étant dit je réalise que dans cette configuration si on va chercher une variable qui n'a pas été envoyée, avant on avait une exception, et maintenant cela va retourner undefined… Il faudra qu'on soit vigilants, à moins que js ait une botte secrète pour raise des exceptions quand même ?

Reply:
> je vois que cette variable n'a pas été reprise, elle n'était jamais utilisée ? (c'est possible…)

Bien vu ! 💪

Effectivement, je n'ai trouvé aucune trace de `serie_periode_annuelle` dans la codebase.

> je réalise que dans cette configuration si on va chercher une variable qui n'a pas été envoyée dans le scope du traitement MR, avant on avait une exception, et maintenant cela va retourner undefined… Il faudra qu'on soit vigilants, à moins que js ait une botte secrète pour raise des exceptions quand même ?

Pas de botte secrète, malheureusement, à ma connaissance...

La stratégie que je propose pour pouvoir intercepter ce genre d'anomalie est de:
- au niveau statique (itérations de développement): typer `jsParams` avec TypeScript => on sera prévenus pendant le développement des fonctions JS si on fait appel à une variable qui n'est pas définie
- au moment de la compilation / CI: mettre en place un test Go qui comparera la structure du `scope` envoyé à MongoDB à celle décrite en TypeScript, pour s'assurer que les deux restent alignés
- à l'exécution: on pourrait écrire une fonction JS (ou la générer à partir des types TypeScript) qui vérifierait que `jsParams` est bien conforme aux attentes + éventuellement une procédure qui ferait une passe de vérification sur le format des données sur le point d'être lues par map-reduce, avant son exécution. (pour éviter des crash à retardement, au bout de plusieurs heures d'exécution)

Ceci dit, je précise qu'aucun de ces dispositifs n'est en place pour l'instant. C'est en cours.